### ISSUE TITLE
Trial feature for simulating missing modules for a test run

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#1795]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#1795|
|Type|enhancement|
|Created|2006-06-06 12:13:53Z|

It would be nice if trial could fake out Python and make it look like a particular module was not present on the system.  ie,

```
trial --without-module OpenSSL twisted
```

would run Twisted tests such that "import OpenSSL" always raised an ImportError.

One implementation technique for this is to put None into sys.modules for the key "OpenSSL".

<details><summary>Searchable metadata</summary>

```
trac-id__1795 1795
type__enhancement enhancement
reporter__exarkun exarkun
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__trial trial
keywords__ 
time__1149596033000000 1149596033000000
changetime__1194376920000000 1194376920000000
version__None None
owner__ 
cc__therve cc__jml
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml commented|
|-|-|

Sounds cool. Maybe it should be part of a more generic "fake platform problems" feature.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set owner to @therve|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_3"></a>|@therve commented|
|-|-|

(In [21495]) KISS

Refs [#1795](https://github.com/twisted/twisted/issues/1795)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_4"></a>|@therve removed owner|
|-|-|

This is ready to review in without-module-1795. I tried to keep the changes minimal.

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_5"></a>|@radix commented|
|-|-|

Here's an idea.

What will happen once I'm using disttrial with ssl connections to my slaves? The current implementation strategy means that it'll break. 

On the other hand, I can't think of any way around this at all, apart from making disttrial have separate processes for network I/O layers and actual tests. bleh, whatever.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve commented|
|-|-|

Replying to [radix](#note_5):

> On the other hand, I can't think of any way around this at all, apart from making disttrial have separate processes for network I/O layers and actual tests. bleh, whatever.

In fact I think that's the way to do it. Because when you use disttrial, you probably also want that remote hosts use multiple processes. Also, that prevents bad side effects like open sockets and all.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set owner to @therve|
|-|-|

test_disableOneModule and test_disableMultipleModules both try to save ```sys.modules``` from potential mangling, but they don't actually manage to: ```savedModules``` is only a reference to the same ```dict``` instance.  Make a copy of ```sys.modules``` at the beginning and then ```clear```/```update``` ```sys.modules``` afterwards (maybe this would be good to do in ```setUp```/```tearDown```).

The lines like ```self.assertIdentical(None, checkSmtp())``` are a bit weird, since those functions always return ```None```.  Maybe they could return what they import and the test could ```assertNotIdentical```?  Although I'm mostly saying that so when you add ```return``` you can delete the ```# pyflakes``` comments which I hate. :)

I wonder if it's worth warning or failing if, in ```opt_without_module```, ```sys.modules``` already has a non-```None``` value for the specified key.  In this case, the module has been imported already and some code, somewhere (who knows what or where), already has a reference to it, so the environment might not behave exactly as if the module were really absent.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_8"></a>|@therve commented|
|-|-|

(In [21610]) Fix some tests, add a warning for already imported modules.

Refs [#1795](https://github.com/twisted/twisted/issues/1795)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_9"></a>|@therve removed owner|
|-|-|

I think I fixed everything, back to review!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun set owner to @therve|
|-|-|

Minor improvement to the man page text:

''
Simulate the lack of the specified comma-separated list of modules.  This makes it look like the modules are not present in the system, causing tests to check the behavior for that configuration.
''

No other issues, merge when you want.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_11"></a>|@therve set status to `closed`|
|-|-|

(In [21618]) Merge without-module-1795

Author: therve
Reviewer: exarkun
Fixes [#1795](https://github.com/twisted/twisted/issues/1795)

Add the --without-module command line option to trial, that simulates the lack
of the specified modules.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
